torch/optim/swa_utils.py

```diff
@@ -0,0 +1,262 @@
+import torch
+import math
+from torch.nn import Module
+from copy import deepcopy
+from torch.optim.lr_scheduler import _LRScheduler
+
+
+class AveragedModel(Module):
+    r"""Implements averaged model for Stochastic Weight Averaging (SWA).
+
+    Stochastic Weight Averaging was proposed in `Averaging Weights Leads to
+    Wider Optima and Better Generalization`_ by Pavel Izmailov, Dmitrii
+    Podoprikhin, Timur Garipov, Dmitry Vetrov and Andrew Gordon Wilson
+    (UAI 2018).
+
+    AveragedModel class creates a copy of the provided module :attr:`model`
+    on the device :attr:`device` and allows to compute running averages of the 
+    parameters of the :attr:`model`.
+
+    Arguments:
+        model (torch.nn.Module): model to use with SWA
+        device (torch.device, optional): if provided, the averaged model will be
+            stored on the :attr:`device` 
+        avg_fn (function, optional): the averaging function used to update 
+            parameters; the function must take in the current value of the 
+            :class:`AveragedModel` parameter, the current value of :attr:`model`
+            parameter and the number of models laready averaged; if None, 
+            equally weighted average is used (default: None)
+
+    Example:
+        >>> loader, optimizer, model, loss_fn = ...
+        >>> swa_model = torch.optim.swa_utils.AveragedModel(model)
+        >>> # swa_model computes an equally-weighted average of the weights
+        >>> # You can use custom averaging functions with `avg_fun` parameter
+        >>> ema_avg = lambda p_avg, p, n_avg: 0.1 * p_avg + 0.9 * p
```

Comment:
Proposal: longer/more verbose lambda arg names in the example code to facilitate matching with the doc language "the function must take in the current value of the 
            :class:`AveragedModel` parameter, the current value of :attr:`model`
            parameter and the number of models laready averaged"

Reply:
I pushed a new version with more explicit names, how does that look?

Reply:
Now looks self-documenting to me, quite nice :) The last thing is less explicit names in the source itself: https://github.com/pytorch/pytorch/pull/35032/files#diff-b0ffccfd1bc4e680950e84b3afdc8ee9R94 (but maybe it's okay)

Reply:
done :)